Add support for tokio-rusqlite

Czocher opened the pull request:
@cljoly please have a look and comment everything that you feel is not right. Consider it a very preliminary version.

There's a few things that are still missing:
- [x] Tests are currently just copied over from the `sync` version to the `async` version. Feels more like doctests should be considered for both versions and the main tests should be performed only on the extracted functions?
- [x] The created feature is not [additive](https://github.com/rust-lang/cargo/issues/4328). I'm not sure how to make it so, maybe providing a `AsyncMigrations` struct and leaving the `Migrations` without any changes (except extracting the functions) may be a solution? What do you thing @cljoly?
- [x] Additional tests can and should be created for the extracted functions in `lib.rs`.

Closes #24.

Czocher commented:
Ah I see you merged the PR to add the up/down hooks. I'll rebase it on the current master later ;).

Czocher commented:
@cljoly I had a far better idea how this could be implemented so that:
1. The changes to the base project are minimized.
2. The changes are additive.

Have a look at: https://github.com/Czocher/rusqlite_migration/pull/new/support-tokio-sqlite-v2

What I did in v2 is to create a simple `AsyncAdapter` holding the base `Migrations` struct in an `Arc`. For an example of the usage of it you can have a look at: https://github.com/Czocher/rusqlite_migration/blob/support-tokio-sqlite-v2/examples/quick_start_async.rs 

Tell me which version you like best (IMHO v2 better handles your requirements as to the implementation).

cljoly commented:
@Czocher thanks for opening this PR, I'll try and take a look on Sunday.

cljoly commented:
Sorry I am very busy this week, I did not get the chance to look into your PR, hopefully this weekend.

Czocher commented:
@cljoly there's no rush :). I suggest you first look into what's proposed in this comment: https://github.com/cljoly/rusqlite_migration/pull/29#issuecomment-1401951224

cljoly reviewed (commented):
I took a look at your v2, it’s much better in terms of code duplication, I quite like it actually! Perhaps you could push the commits from the v2 branch into this branch, to ease further review?

Just a minor point, we could rename `AsyncAdapter` to `AsyncMigration` there.

cljoly commented:
>  Tests are currently just copied over from the sync version to the async version. Feels more like doctests should be considered for both versions and the main tests should be performed only on the extracted functions?

Yes, but this is much less of a problem in your v2 branch :)

>  Additional tests can and should be created for the extracted functions in lib.rs.

We want to have good test coverage indeed 👍🏼

cljoly commented:
>  we should discuss if Migrations are ever used with a lifetime 'm which is not 'static. If Migrations are always 'static it would simplify the async version slightly.

Right, I think there are cases where you don’t use 'static. In any case, I believe going to a more restrictive lifetime parameter would be a breaking API change, so that should be avoided if possible.

cljoly commented:
I appreciate all the work you have put in this @Czocher  and sorry I could not review earlier. Hopefully, I have addressed all your questions (that were not solved in v2).

cljoly commented:
> Ah I see you merged the PR to add the up/down hooks. I'll rebase it on the current master later ;).

That’s fine and I suspect you will have less conflicts (if at all in the v2 branch)

cljoly commented on src/async_tokio_rusqlite.rs:
````diff
@@ -0,0 +1,328 @@
+    ///     M::up("CREATE TABLE food (name TEXT);"),
+    /// ]);
+    /// ```
+    pub fn new(ms: Vec<M>) -> Self {
````
Actually, that’s irrelevant, let’s ignore that comment.

Czocher commented on Cargo.toml:
```diff
@@ -9,21 +9,27 @@ keywords = ["rusqlite", "sqlite", "user_version", "database", "migration"]
 categories = ["database"]
 readme = "README.md"
 homepage = "https://cj.rs/rusqlite_migration"
+repository = "https://github.com/cljoly/rusqlite_migration"
```
Sure, I see you're very strict about atomic PRs, which I appreciate a lot - I have the same approach in the projects I control :). I'll make every PR atomic from now on. 

This change has been extracted to #30.

cljoly replied:
Great thank you!

cljoly replied:
(and I’m by no means perfect, I sometimes mix up multiple changes as well :) )

Czocher commented on Cargo.toml:
```diff
@@ -9,21 +9,27 @@ keywords = ["rusqlite", "sqlite", "user_version", "database", "migration"]
 [dev-dependencies]
 simple-logging = "2.0.2"
-env_logger = "0.9"
+env_logger = "0.10.0"
```
Extracted to #31.

Czocher commented:
Replaced the branch in this PR with the the `v2` branch. No need to review yet, the only change I implemented is the rename from `AsyncAdapter` to `AsyncMigration` (shouldn't it be named `AsyncMigrations` instead?).

What needs to be done before this change should be reviewed:
- [ ] Extract the `AsyncMigration` to an `asynch` submodule.
- [ ] Add documentation.
- [ ] Add unit tests.
- [ ] Update the `README.md` to indicate that `tokio-rusqlite` is supported and point to the example.

cljoly commented on src/lib.rs:
```diff
@@ -91,6 +91,9 @@ use log::{debug, info, trace, warn};
 
+/// Support for asynchonous migrations using tokio-rusqlite.
+#[cfg(feature = "async-tokio-rusqlite")]
+pub mod asynch;
```
Perhaps it would be more ergonomic not to expose this module publicly, just the `AsyncMigrations` struct, what do you think?

Czocher replied:
I have no strong feelings :).

cljoly replied:
Hum, then I prefer having the module private 😃

cljoly commented on Cargo.toml:
```diff
@@ -11,9 +11,14 @@ readme = "README.md"
 [dependencies]
+tokio = { version = "1.24.2", features = ["full"], optional = true }
+tokio-rusqlite = { version = "0.3.0", optional = true }
+tokio-test = { version = "0.4.2", optional = true }
```
That can be moved to `dev-dependencies`:
```diff
diff --git a/Cargo.toml b/Cargo.toml
index b75b5cb..c3e9e45 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -13,12 +13,11 @@ rust-version = "1.61"
 
 [features]
 default = []
-async-tokio-rusqlite = ["dep:tokio-rusqlite", "dep:tokio", "dep:tokio-test"]
+async-tokio-rusqlite = ["dep:tokio-rusqlite", "dep:tokio"]
 
 [dependencies]
 tokio = { version = "1.24.2", features = ["full"], optional = true }
 tokio-rusqlite = { version = "0.3.0", optional = true }
-tokio-test = { version = "0.4.2", optional = true }
 log = "0.4"
 
 [dependencies.rusqlite]
@@ -27,6 +26,8 @@ default-features = false
 features = []
 
 [dev-dependencies]
 simple-logging = "2.0.2"
 env_logger = "0.9"
 anyhow = "1"
+tokio-test = { version = "0.4.2" }
```

Czocher replied:
Done :).

cljoly reviewed (commented):
I’ll look more into this later, but here are some first few comments.

cljoly commented:
I need to think a bit more about unit tests and how to disable the async example by default.

And also,
- [ ] the CI would need to be changed to test with and without the feature (I can take care of it if you want)

Czocher commented:
> I need to think a bit more about unit tests and how to disable the async example by default.

From what I saw usually people make a `cargo` `workspace` in the main package directory and then make the examples separate projects which include the features they require.

cljoly commented:
> From what I saw usually people make a cargo workspace in the main package directory and then make the examples separate projects which include the features they require.

Since I have not found a better option, let’s do that if you don’t mind.

> Check if I did it correctly ;). I added a test with all-features to future-proof against non-additive features which may be happen at some point.

The CI looks great, thanks!

Czocher commented on README.md:
````diff
@@ -81,6 +81,7 @@ conn.execute("INSERT INTO friend (name) VALUES (?1)", params!["John"])
 ```
 
 Please see the [examples](https://github.com/cljoly/rusqlite_migrate/tree/master/examples) folder for more, in particular:
+- `async` migrations in the (quick_start_async.rs)[https://github.com/cljoly/rusqlite_migrate/tree/master/examples/quick_start_async.rs)] file
````
Ah, so I'll need to add it when the file exists in the `master` branch?

cljoly replied:
No sorry if that was unclear.

I'm using [a tool](https://github.com/phaazon/cargo-sync-readme/) to copy the content of the main documentation comment in lib.rs to a part of the reasme file delimited by comments. The CI checks that the content of rlthe readme and of lib.rs matches, that's why it currently fails.

So you need to install https://github.com/phaazon/cargo-sync-readme/, make the change in lib.rs and run it to apply the changes to the readme, in the PR. A hacky way would be to just copy your changes though 🙂

I hope that makes sense, I need to document that better

Czocher commented:
> Since I have not found a better option, let’s do that if you don’t mind.

I implemented the structure change here since it's required for this PR to compile. If you feel like it should be extracted to a separate PR let me now ;).

Czocher commented:
In my opinion this task is mergeable at the moment, only thing potentially missing are unit tests for the async version. I wrote an `async_integration_test` to do some basic checks.

cljoly commented:
> In my opinion this task is mergeable at the moment

Great! I should be able to take a deeper look this weekend and if everything goes well, to merge.

cljoly commented:
I think it looks great, and this can be further refined if needed in the future. I’m sorry I introduced a conflict when merging another PR, I’ll solve the conflict when I get the chance and merge.

cljoly commented on examples/simple/src/main.rs:
```diff
@@ -18,7 +18,7 @@ mod tests {
     static ref MIGRATIONS: Migrations<'static> =
         Migrations::new(vec![
-            M::up(include_str!("friend_car.sql")),
+            M::up(include_str!("../../friend_car.sql")),
```
While I’m rebasing, I’ll move that file in the simple example folder.

cljoly replied:
Hum on a second thought, it’s used in integration tests as well, let’s leave it there.

cljoly commented on rusqlite_migration_tests/tests/integration_test.rs:
```diff
@@ -11,7 +11,7 @@ fn main_test() {
     let mut ms = vec![
         M::up("CREATE TABLE t(a);"),
-        M::up(include_str!("../examples/friend_car.sql")),
+        M::up(include_str!("../../examples/friend_car.sql")),
```
here too

cljoly commented:
Alright, I’ve also fixed a markdown link :)

cljoly commented:
Thank you so much for your contribution and your patience!
